Gradle plugin (1.4.4) regression: NPE while executing task: bloopInstall

Thanks for providing a Gradle plugin for bloop!

We use it for generating bloop config, so that we enjoy super fast compilation in the IDE.
We apply the plugin like:
```      
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {  "classpath"(group = "ch.epfl.scala", name = "gradle-bloop_2.12", version = "1.4.3")}
}
plugins {
    // Apply the scala plugin to add support for Scala
    scala

    // Apply the java-library plugin for API and implementation separation.
    `java-library`

    // To easily launch
    application
}

apply(plugin = "bloop")

application {
    mainClass.set("com.foo.bar.Main")
}
```
But when trying to update to version 1.4.4, we get the following NPE when executing `./gradlew bloopInstall`:
```
Caused by: java.lang.NullPointerException
        at com.github.plokhotnyuk.jsoniter_scala.core.JsonWriter.writeVal(JsonWriter.scala:325)
        at bloop.config.ConfigCodecs$$anon$8$$anon$10.e0(ConfigCodecs.scala:133)
        at bloop.config.ConfigCodecs$$anon$8$$anon$10.encodeValue(ConfigCodecs.scala:133)
        at bloop.config.ConfigCodecs$$anon$8$$anon$10.encodeValue(ConfigCodecs.scala:133)
        at bloop.config.ConfigCodecs$$anon$8.encodeValue(ConfigCodecs.scala:136)
        at bloop.config.ConfigCodecs$$anon$8.encodeValue(ConfigCodecs.scala:128)
        at bloop.config.ConfigCodecs$$anon$11$$anon$12.e1(ConfigCodecs.scala:171)
        at bloop.config.ConfigCodecs$$anon$11$$anon$12.e0(ConfigCodecs.scala:171)
        at bloop.config.ConfigCodecs$$anon$11$$anon$12.encodeValue(ConfigCodecs.scala:171)
        at bloop.config.ConfigCodecs$$anon$11$$anon$12.encodeValue(ConfigCodecs.scala:171)
        at bloop.config.ConfigCodecs$$anon$11.encodeValue(ConfigCodecs.scala:181)
        at bloop.config.ConfigCodecs$$anon$11.encodeValue(ConfigCodecs.scala:168)
        at bloop.config.ConfigCodecs$$anon$13.e0(ConfigCodecs.scala:200)
        at bloop.config.ConfigCodecs$$anon$13.encodeValue(ConfigCodecs.scala:200)
        at bloop.config.ConfigCodecs$$anon$13.encodeValue(ConfigCodecs.scala:200)
        at bloop.config.ConfigCodecs$$anon$14.e0(ConfigCodecs.scala:203)
        at bloop.config.ConfigCodecs$$anon$14.encodeValue(ConfigCodecs.scala:203)
        at bloop.config.ConfigCodecs$$anon$14.encodeValue(ConfigCodecs.scala:203)
        at com.github.plokhotnyuk.jsoniter_scala.core.JsonWriter.write(JsonWriter.scala:587)
        at com.github.plokhotnyuk.jsoniter_scala.core.package$.writeToArray(package.scala:208)
        at bloop.config.ConfigCodecs$.toStr(ConfigCodecs.scala:218)
        at bloop.config.package$.write(package.scala:9)
        at bloop.config.package$.write(package.scala:11)
        at bloop.integrations.gradle.tasks.BloopInstallTask.generateBloopConfiguration(BloopInstallTask.scala:90)
        at bloop.integrations.gradle.tasks.BloopInstallTask.runBloopPlugin(BloopInstallTask.scala:52)
        at bloop.integrations.gradle.tasks.BloopInstallTask.run(BloopInstallTask.scala:30)
```

Without appling the application gradle plugin, `./gradlew bloopInstall` seems to work as expected.
So it looks like an incompatibility with the application plugin?

--- a/integrations/gradle-bloop/src/main/scala/bloop/integrations/gradle/model/BloopConverter.scala
+++ b/integrations/gradle-bloop/src/main/scala/bloop/integrations/gradle/model/BloopConverter.scala
@@ -21,7 +21,6 @@ import org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact
 import org.gradle.api.internal.artifacts.publish.{ArchivePublishArtifact, DecoratingPublishArtifact}
 import org.gradle.api.internal.file.copy.DefaultCopySpec
 import org.gradle.api.internal.tasks.compile.{DefaultJavaCompileSpec, JavaCompilerArgumentsBuilder}
-import org.gradle.api.plugins.{ApplicationPluginConvention, JavaPluginConvention}
 import org.gradle.api.tasks.bundling.AbstractArchiveTask
 import org.gradle.api.tasks.{AbstractCopyTask, SourceSet}
 import org.gradle.api.tasks.compile.{CompileOptions, JavaCompile}
@@ -324,7 +323,6 @@ class BloopConverter(parameters: BloopParameters) {
 
   private def getSourceSetProjectMap(rootProject: Project): Map[SourceSet, Project] = {
     rootProject.getAllprojects.asScala
-      .filter(isJavaProject)
       .flatMap(p => p.allSourceSets.map(_ -> p))
       .toMap
   }
@@ -373,7 +371,6 @@ class BloopConverter(parameters: BloopParameters) {
       sourceSets: Set[SourceSet]
   ): Map[File, SourceSet] = {
     rootProject.getAllprojects.asScala
-      .filter(isJavaProject)
       .flatMap(_.getConfigurations.asScala.flatMap(c => {
         val archiveTasks = c.getAllArtifacts.asScala.flatMap(getArchiveTask)
         val possibleArchiveSourceSets =
@@ -418,6 +415,10 @@ class BloopConverter(parameters: BloopParameters) {
           task.getJvmArgs.asScala.toList ++
           testProperties
       })
+      .orElse(
+        project.javaApplicationExt
+          .flatMap(f => Option(f.getApplicationDefaultJvmArgs).map(_.asScala.toList))
+      )
       .getOrElse(
         Option(forkOptions.getMemoryInitialSize).map(mem => s"-Xms$mem").toList ++
           Option(forkOptions.getMemoryMaximumSize).map(mem => s"-Xmx$mem").toList ++
@@ -426,8 +427,7 @@ class BloopConverter(parameters: BloopParameters) {
 
     val mainClass =
       if (testTask.isEmpty)
-        Option(project.getConvention.findPlugin(classOf[ApplicationPluginConvention]))
-          .map(_.getMainClassName)
+        project.javaApplicationExt.flatMap(f => Option(f.getMainClassName))
       else
         None
 
@@ -494,10 +494,6 @@ class BloopConverter(parameters: BloopParameters) {
     } catch {
       case _: NoSuchMethodException => None;
     }
-  }
-
-  private def isJavaProject(project: Project): Boolean = {
-    project.getConvention.findPlugin(classOf[JavaPluginConvention]) != null
   }
 
   private def getOutDir(targetDir: File, projectName: String): Path =

--- a/integrations/gradle-bloop/src/main/scala/bloop/integrations/gradle/syntax.scala
+++ b/integrations/gradle-bloop/src/main/scala/bloop/integrations/gradle/syntax.scala
@@ -3,7 +3,8 @@ package bloop.integrations.gradle
 import java.io.File
 
 import org.gradle.api.artifacts.Configuration
-import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.plugins.JavaApplication
 import org.gradle.api.tasks.SourceSet
 import org.gradle.api.{Project, Task, GradleException}
 
@@ -30,21 +31,14 @@ object syntax {
       configs.findByName(name)
     }
 
-    def getSourceSet(name: String): SourceSet = {
-      val plugins = project.getConvention.getPlugins
-      try {
-        project.getConvention.getPlugin(classOf[JavaPluginConvention]).getSourceSets.getByName(name)
-      } catch {
-        case NonFatal(e) =>
-          throw new GradleException(
-            s"Could not find java plugin convention for $name in $project with plugins $plugins",
-            e
-          )
-      }
+    def allSourceSets: Set[SourceSet] = {
+      Option(project.getExtensions.findByType(classOf[SourceSetContainer]))
+        .map(_.asScala.toSet)
+        .getOrElse(Set.empty)
     }
 
-    def allSourceSets: Set[SourceSet] = {
-      project.getConvention.getPlugin(classOf[JavaPluginConvention]).getSourceSets.asScala.toSet
+    def javaApplicationExt: Option[JavaApplication] = {
+      Option(project.getExtensions.findByType(classOf[JavaApplication]))
     }
 
     def createExtension[T](name: String, params: Object*)(implicit t: ClassTag[T]): Unit = {

--- a/integrations/gradle-bloop/src/test/scala/bloop/integrations/gradle/ConfigGenerationSuite.scala
+++ b/integrations/gradle-bloop/src/test/scala/bloop/integrations/gradle/ConfigGenerationSuite.scala
@@ -550,7 +550,6 @@ abstract class ConfigGenerationSuite {
       .withProjectDir(testProjectDir.getRoot)
       .withPluginClasspath(getClasspath)
       .withArguments("bloopInstall", "-Si")
-      .withDebug(true)
       .build()
 
     val projectName = testProjectDir.getRoot.getName
@@ -1572,6 +1571,103 @@ abstract class ConfigGenerationSuite {
          |
          |mainClassName = 'org.main.name'
          |
+         |application {
+         |  applicationDefaultJvmArgs = ["-Dgreeting.language=en", "-Xmx16g"]
+         |}
+      """.stripMargin
+    )
+
+    createHelloWorldJavaSource(testProjectDir.getRoot)
+    createHelloWorldJavaTestSource(testProjectDir.getRoot)
+
+    GradleRunner
+      .create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.getRoot)
+      .withPluginClasspath(getClasspath)
+      .withArguments("bloopInstall", "-Si")
+      .build()
+
+    val projectName = testProjectDir.getRoot.getName
+    val bloopDir = new File(testProjectDir.getRoot, ".bloop")
+    val projectFile = new File(bloopDir, s"${projectName}.json")
+    val projectTestFile = new File(bloopDir, s"${projectName}-test.json")
+    val projectConfig = readValidBloopConfig(projectFile)
+    assert(projectConfig.project.`scala`.isEmpty)
+    assert(projectConfig.project.platform.nonEmpty)
+    assert(projectConfig.project.platform.get.mainClass.nonEmpty)
+    assertEquals("org.main.name", projectConfig.project.platform.get.mainClass.get)
+    val platform = projectConfig.project.platform
+    val config = platform.get.asInstanceOf[Platform.Jvm].config
+    assertEquals(Set("-Dgreeting.language=en", "-Xmx16g"), config.options.toSet)
+    assert(projectConfig.project.dependencies.isEmpty)
+    assert(projectConfig.project.classpath.isEmpty)
+    assert(hasTag(projectConfig, Tag.Library))
+
+    val projectTestConfig = readValidBloopConfig(projectTestFile)
+    assert(projectConfig.project.`scala`.isEmpty)
+    assertEquals(projectTestConfig.project.dependencies, List(projectName))
+    assert(hasTag(projectTestConfig, Tag.Test))
+    assert(compileBloopProject(s"${projectName}-test", bloopDir).status.isOk)
+  }
+
+  @Test def generateConfigFileForOtherMainClass(): Unit = {
+    if (gradleVersion >= "6.4") {
+      val buildFile = testProjectDir.newFile("build.gradle")
+      writeBuildScript(
+        buildFile,
+        s"""
+           |plugins {
+           |  id 'bloop'
+           |}
+           |apply plugin: 'application'
+           |apply plugin: 'java'
+           |apply plugin: 'bloop'
+           |
+           |application {
+           |  mainClass.set("org.main.name")
+           |  applicationDefaultJvmArgs = ["-Dgreeting.language=en", "-Xmx16g"]
+           |}
+           |
+      """.stripMargin
+      )
+
+      createHelloWorldJavaSource(testProjectDir.getRoot)
+
+      GradleRunner
+        .create()
+        .withGradleVersion(gradleVersion)
+        .withProjectDir(testProjectDir.getRoot)
+        .withPluginClasspath(getClasspath)
+        .withArguments("bloopInstall", "-Si")
+        .build()
+
+      val projectName = testProjectDir.getRoot.getName
+      val bloopDir = new File(testProjectDir.getRoot, ".bloop")
+      val projectFile = new File(bloopDir, s"${projectName}.json")
+      val projectTestFile = new File(bloopDir, s"${projectName}-test.json")
+      val projectConfig = readValidBloopConfig(projectFile)
+      assert(projectConfig.project.`scala`.isEmpty)
+      val platform = projectConfig.project.platform
+      assert(platform.isDefined)
+      assert(platform.get.isInstanceOf[Platform.Jvm])
+      assert(platform.get.mainClass.isDefined)
+      assertEquals("org.main.name", platform.get.mainClass.get)
+    }
+  }
+
+  @Test def generateConfigFileForNullMainClass(): Unit = {
+    val buildFile = testProjectDir.newFile("build.gradle")
+    writeBuildScript(
+      buildFile,
+      s"""
+         |plugins {
+         |  id 'bloop'
+         |}
+         |apply plugin: 'application'
+         |apply plugin: 'java'
+         |apply plugin: 'bloop'
+         |
       """.stripMargin
     )
 
@@ -1598,7 +1694,9 @@ abstract class ConfigGenerationSuite {
 
     val projectTestConfig = readValidBloopConfig(projectTestFile)
     assert(projectConfig.project.`scala`.isEmpty)
-    assert(projectTestConfig.project.dependencies == List(projectName))
+    assert(projectConfig.project.platform.nonEmpty)
+    assert(projectConfig.project.platform.get.mainClass.isEmpty)
+    assertEquals(projectTestConfig.project.dependencies, List(projectName))
     assert(hasTag(projectTestConfig, Tag.Test))
     assert(compileBloopProject(s"${projectName}-test", bloopDir).status.isOk)
   }
@@ -1644,7 +1742,7 @@ abstract class ConfigGenerationSuite {
     assert(platform.get.isInstanceOf[Platform.Jvm])
     val config = platform.get.asInstanceOf[Platform.Jvm].config
     if (!scala.util.Properties.isWin) assert(config.home.contains(Paths.get("/opt/jdk11")))
-    assert(config.options.toSet == Set("-XX:MaxMetaSpaceSize=512m", "-Xms1g", "-Xmx2g"))
+    assertEquals(Set("-XX:MaxMetaSpaceSize=512m", "-Xms1g", "-Xmx2g"), config.options.toSet)
   }
 
   @Test def importsTestSystemProperties(): Unit = {
@@ -1686,14 +1784,15 @@ abstract class ConfigGenerationSuite {
     assert(platform.isDefined)
     assert(platform.get.isInstanceOf[Platform.Jvm])
     val config = platform.get.asInstanceOf[Platform.Jvm].config
-    assert(
-      config.options.toSet == Set(
+    assertEquals(
+      Set(
         "-XX:+UseG1GC",
         "-verbose:gc",
         "-Xms1g",
         "-Xmx2g",
         "-Dproperty=value"
-      )
+      ),
+      config.options.toSet
     )
   }
 
@@ -1745,7 +1844,7 @@ abstract class ConfigGenerationSuite {
     )
 
     val setupA = getSetup(buildDirA)
-    assert(setupA.order == JavaThenScala)
+    assertEquals(JavaThenScala, setupA.order)
 
     val buildDirB = testProjectDir.newFolder("b")
     val buildFileB = new File(buildDirB, "build.gradle")
@@ -1777,7 +1876,7 @@ abstract class ConfigGenerationSuite {
     )
 
     val setupB = getSetup(buildDirB)
-    assert(setupB.order == Mixed)
+    assertEquals(Mixed, setupB.order)
   }
 
   @Test def maintainsClassPathOrder(): Unit = {
@@ -2312,7 +2411,7 @@ abstract class ConfigGenerationSuite {
     assert(configFile.project.dependencies.isEmpty)
     assert(hasTag(configFile, Tag.Library))
 
-    assert(configTestFile.project.dependencies == List(projectName))
+    assertEquals(List(projectName), configTestFile.project.dependencies)
     assert(hasTag(configTestFile, Tag.Test))
     assert(compileBloopProject(s"${projectName}-test", bloopDir).status.isOk)
     assertAllConfigsMatchJarNames(List(configFile, configTestFile), List("scala-library"))

